packages/cursorless-vscode/src/CommandHistory.ts

```diff
@@ -0,0 +1,147 @@
+import {
+  ActionDescriptor,
+  CommandComplete,
+  CommandHistoryItem,
+  Disposable,
+  FileSystem,
+  ReadOnlyHatMap,
+} from "@cursorless/common";
+import type {
+  CommandRunner,
+  CommandRunnerDecorator,
+} from "@cursorless/cursorless-engine";
+import * as fs from "node:fs/promises";
+import * as path from "node:path";
+import * as vscode from "vscode";
+
+const filePrefix = "cursorlessCommandHistory";
+const settingSection = "cursorless";
+const settingName = "commandHistory";
+const settingFullName = `${settingSection}.${settingName}`;
+
+export class CommandHistory implements CommandRunnerDecorator {
+  private readonly dirPath: string;
+  private readonly cursorlessVersion: string;
+  private disposable: Disposable;
+  private isActive: boolean = false;
+
+  constructor(
+    extensionContext: vscode.ExtensionContext,
+    fileSystem: FileSystem,
+  ) {
+    this.cursorlessVersion = extensionContext.extension.packageJSON.version;
+    this.dirPath = fileSystem.cursorlessCommandHistoryDirPath;
+
+    // Read initial setting value. The watcher below will take care of changes.
+    this.evaluateSetting();
+
+    this.disposable = vscode.workspace.onDidChangeConfiguration((event) => {
+      if (event.affectsConfiguration(settingFullName)) {
+        this.evaluateSetting();
+      }
+    });
+  }
+
+  wrapCommandRunner(
+    readableHatMap: ReadOnlyHatMap,
+    runner: CommandRunner,
+  ): CommandRunner {
+    if (!this.isActive) {
+      return runner;
+    }
+
+    return {
+      run: async (commandComplete: CommandComplete) => {
```

Comment:
Should we capture the errors? Might be useful to know how often people are running into inference issues, etc. I think just the error name would suffice. Worth quick look at how `TestCaseRecorder` wraps command in case helpful https://github.com/cursorless-dev/cursorless/blob/40a6fee4c4eec9b37cfb1ee7b4bf5ff78df59ea2/packages/cursorless-engine/src/testCaseRecorder/TestCaseRecorder.ts#L458-L483

Reply:
But we are not logging the result? Should we log the command itself and then optionally log errors?

Reply:
Not sure I follow. You could just add an optional error field to the top-level payload

Reply:
I just have my mind set on logging before executing the command. I guess we could add it afterwards.
Updated